src/subscription/subscription.service.ts

```diff
@@ -1,4 +1,296 @@
-import { Injectable } from '@nestjs/common';
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { AnalyticsService } from 'src/analytics/analytics.service';
+import { ApiRequestDto } from './dto/make-request.dto';
+import { ZaLaResponse } from 'src/common/helpers/response';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Api } from 'src/entities/api.entity';
+import { Profile } from 'src/entities/profile.entity';
+import { Endpoint } from 'src/entities/endpoint.entity';
+import { HttpService } from '@nestjs/axios';
+import { JwtService } from '@nestjs/jwt';
+import { Repository } from 'typeorm';
+import { Subscription } from 'src/entities/subscription.entity';
+import { CreateSubDto } from './dto/create-subscription.dto';
+import { Tokens } from 'src/common/interfaces/subcriptionToken.interface';
 
 @Injectable()
-export class SubscriptionService {}
+export class SubscriptionService {
+  constructor(
+    private analyticsService: AnalyticsService,
+    @InjectRepository(Subscription)
+    private readonly subRepository: Repository<Subscription>,
+    @InjectRepository(Api)
+    private readonly apiRepository: Repository<Api>,
+    @InjectRepository(Profile)
+    private readonly profileRepository: Repository<Profile>,
+    @InjectRepository(Endpoint)
+    private readonly endpointRepository: Repository<Endpoint>,
+    private httpService: HttpService,
+    private jwtService: JwtService,
+  ) {}
+
+  /**
+   * It takes a token and a body as parameters, verifies the token, finds the endpoint, makes a request
+   * to the endpoint, and returns the response.
+   * @param {string} token - the token that is generated for the user when they subscribe to an api
+   * @param {ApiRequestDto} body - ApiRequestDto
+   * @returns The data from the external api.
+   */
+  async apiRequest(token: string, body: ApiRequestDto): Promise<any> {
+    const { api, profile } = await this.verifySub(token);
+    const encodedRoute = encodeURIComponent(body.route);
```

Comment:
on creating this endpoint in the endpoint service.. the route is not encoded.. so this might not work.. encode the route so this would work

Reply:
On creation, route is encoded in the endpoint entity, I used encodeURIcomponent(), a built in Javascript function

Reply:
oh... i can see it.. nice work